templates/web/pages/product_edit/product_edit_form_packagings.tt.html

```diff
@@ -0,0 +1,182 @@
+var packaging_materials_select2_options = [% generate_select2_options_for_taxonomy_to_json("packaging_materials") %];
+var packaging_recycling_select2_options = [% generate_select2_options_for_taxonomy_to_json("packaging_recycling") %];
+
+var component = 0;
+</initjs>
+
+<p>[% lang('packaging_parts') %]</p>
+    
+<div id="packagings_components_edit">
+
+    <!-- header row for large display -->
+    <div class="row show-for-large-up" id="packagings_components_edit_header">
+        <div class="large-1 columns">[% lang("packaging_number_of_units") %]</div>
+        <div class="large-3 columns">[% lang("packaging_shape") %]</div>
+        <div class="large-3 columns">[% lang("packaging_material") %]</div>
+        <div class="large-3 columns">[% lang("packaging_recycling") %]</div>
+        <div class="large-1 columns">[% lang("packaging_weight") %] (g)</div>
+        <div class="large-1 columns">[% lang("packaging_quantity_per_unit") %]</div>
+    </div>
+
+[% FOREACH packaging IN product.packagings %]
+    [% SET packaging_max = loop.index() %]
+    [% SET packaging_id = "packaging_" _ loop.index() %]
+    <div class="row packagings_components_edit_row" id="[% packaging_id %]_row">
+        <div class="small-12 large-1 columns">
+
+            <label class="hide-for-large-up" for="[% packaging_id %]_number_of_units">[% lang("packaging_number_of_units") %]</label>
+
+            <div class="row">
+                <div class="small-12 large-4 columns delete_packaging_column">
+                    <a class="delete_packaging" id="[% packaging_id %]_delete">
+                        <span class="material-icons">
+                            delete
+                        </span>
+                    </a>
+                </div>
+
+                <div class="small-12 large-8 columns">
+                    <input id="[% packaging_id %]_number_of_units" name="[% packaging_id %]_number_of_units" type="text" value="[% packaging.number_of_units %]">
+                </div>
+            </div>
+        </div>
+
+        [% FOREACH property IN ["shape", "material", "recycling"] %]
+
+            [% IF property == "shape" %]
+                [% SET taxonomy = "packaging_shapes" %]
+            [% ELSIF property == "material" %]
+                [% SET taxonomy = "packaging_materials" %]
+            [% ELSIF property == "recycling" %]
+                [% SET taxonomy = "packaging_recycling" %]
+            [% END %]
+
+            [% SET select2_id = packaging_id _ "_$property" %]
+
+        <div class="small-12 large-3 columns">
+            <label class="hide-for-large-up" for="[% select2_id %]">[% lang("packaging_$property") %]</label>
+            <select id="[% select2_id %]" name="[% select2_id %]" class="form-control [% taxonomy %]_select2">
+                <option></option>
+            </select>
+        </div>
+
+    [% IF packaging_max > 0 %]
```

Comment:
This condition is obscure to me ! Is it to test if we have a value ?

Reply:
```suggestion
    [%# row 0 is a template for new row, we do not populate it %]
    [% IF packaging_max > 0 %]
```

But in this case does it means you have an empty first row in product.packagings ?

Reply:
product_multilingual.pl adds an empty row at the beginning and at the end:

```
	# Add an empty packaging element to the form, that will be hidden and duplicated when the user adds new packaging items,
	# and another empty packaging element at the end
	if (not defined $product_ref->{packagings}) {
		$product_ref->{packagings} = [];
	}
	my $number_of_packaging_components = scalar @{$product_ref->{packagings}};

	unshift(@{$product_ref->{packagings}}, {});
	push(@{$product_ref->{packagings}}, {});
```

As it's the first empty row, it's hidden, so there's no point in creating select2 on it.

Then when product_multilingual.pl reads back the values, it ignores the empty rows:

```
	# And then we add each packaging component
	for (my $packaging_id = 1; $packaging_id <= single_param("packaging_max"); $packaging_id++) {
```

I will add a clearer comment.

Reply:
yes we need a good comment :-)